bundles/org.openhab.binding.boschshc/src/main/java/org/openhab/binding/boschshc/internal/devices/bridge/BoschHttpClient.java

```diff
@@ -0,0 +1,158 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.boschshc.internal.devices.bridge;
+
+import static org.eclipse.jetty.http.HttpMethod.GET;
+
+import java.nio.charset.StandardCharsets;
+import java.security.KeyStoreException;
+import java.security.cert.Certificate;
+import java.security.cert.CertificateEncodingException;
+import java.util.Base64;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeoutException;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jetty.client.HttpClient;
+import org.eclipse.jetty.client.api.ContentResponse;
+import org.eclipse.jetty.client.util.StringContentProvider;
+import org.eclipse.jetty.util.ssl.SslContextFactory;
+import org.openhab.binding.boschshc.internal.exceptions.PairingFailedException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+
+/**
+ * HTTP client using own context with private & Bosch Certs
+ * to pair and connect to the Bosch Smart Home Controller.
+ *
+ * @author Gerd Zanker - Initial contribution
+ */
+@NonNullByDefault
+public class BoschHttpClient extends HttpClient {
+
+    private static final int MAX_PAIR_TRIES = 12;
+
+    private final Logger logger = LoggerFactory.getLogger(BoschHttpClient.class);
+
+    private String ipAddress;
+    private String systempassword;
+
+    public BoschHttpClient(String ipAddress, String systempassword, SslContextFactory sslContextFactory) {
+        super(sslContextFactory);
+        this.ipAddress = ipAddress;
+        this.systempassword = systempassword;
+    }
+
+    private String getCertFromSslContextFactory() throws KeyStoreException, CertificateEncodingException {
+        Certificate cert = this.getSslContextFactory().getKeyStore().getCertificate(BoschSslUtil.getBoschSHCId());
+        return new String(Base64.getEncoder().encode(cert.getEncoded()));
+    }
+
+    public void checkAccessAndPairIfNecessary() throws InterruptedException, PairingFailedException {
+        // test if pairing is needed
+        int counter = 0;
+        boolean accessPossible = isAccessPossible();
+        while (!accessPossible) {
+            // sleep some seconds after every try, except the first one
+            if (counter > 0) {
+                logger.info("Last Pairing failed, starting retry number {}/{} in few seconds", counter, MAX_PAIR_TRIES);
```

Comment:
To add to this. In general the user should not have to look into the log file in case of problems, but this should be reflected in the thing status. Only when there are real problems, like a possible bug, the process is to enable debug log level to get detailed information a user can use to get an idea what the problem might be.

Reply:
@GerdZanker changed this to `logger.trace`